### ISSUE TITLE
ERROR GetMonitoredDatabasesFromConfigDB

### ISSUE THREAD

**Issue by yinan8128** [cybertec-postgresql/pgwatch2#359]
Hi pgwatch2 team,

Once we config pgwatch2.monitored_db md_port to 6432 to use pgbouncer, have error "ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist", as the understanding because pgwatch2 service still use postgresql 5432 port, should not have this error, Please help check, thanks

pgwatch2 version: 1.8.3
OS: CentOS Linux release 7.6.1810
pgbouncer: 1.14.0
postgresql: 11.11

systemctl status pgwatch2 -l
● pgwatch2.service - PGWatch2
   Loaded: loaded (/etc/systemd/system/pgwatch2.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/pgwatch2.service.d
           └─override.conf, required.conf
   Active: active (running) since Thu 2021-03-25 09:04:25 UTC; 22h ago
 Main PID: 12752 (pgwatch2-daemon)
   CGroup: /system.slice/pgwatch2.service
           └─12752 /usr/bin/pgwatch2-daemon --datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1 --dbname=postgres --password=

Mar 25 09:04:25 s34567y-patroni_cluster-932280702.acronis systemd[1]: Starting PGWatch2...
Mar 25 09:04:25 s34567y-patroni_cluster-932280702.acronis systemd[1]: Started PGWatch2.
Mar 25 20:32:30 s34567y-patroni_cluster-932280702.acronis pgwatch2[12752]: 2021/03/25 20:32:30 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
You have new mail in /var/spool/mail/root

(1). pgwatch2.monitored_db record

postgres=# select md_unique_name, md_hostname, md_port, md_dbname, md_user, md_preset_config_name, md_dbtype, md_include_pattern, md_exclude_pattern, md_host_config from pgwatch2.monitored_db;
-[ RECORD 1 ]---------+---------------------------------------------------------------------------------------------------------------------
md_unique_name        | patroni
md_hostname           | 127.0.0.1
md_port               | 6432
md_dbname             |
md_user               | pgwatch2
md_preset_config_name | exhaustive
md_dbtype             | postgres-continuous-discovery
md_include_pattern    | .*
md_exclude_pattern    | (template0|template1)
md_host_config        | {"scope": "pgcluster11", "dcs_type": "consul", "namespace": "/service/", "dcs_endpoints": ["http://localhost:8500"]}

(2). /etc/sysconfig/pgwatch2.conf
OPTIONS='--datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1 --dbname=postgres --password='

(3). lsof -i :5432
pgwatch2- 12752 prometheus    3u  IPv4 592882059      0t0  TCP localhost:46942->localhost:postgres (ESTABLISHED)
...

**Comment by kmoppel** [comment]
Seems to be duplicate of https://github.com/cybertec-postgresql/pgwatch2/issues/348

**Comment by yinan8128** [comment]
@kmoppel 

This is not a duplicate of #348

#348 is both pgwatch2 service port and pgwatch2.monitored_db md_port config to 6432

This ticket is only pgwatch2.monitored_db md_port config to 6432, as the understanding because pgwatch2 service still use postgresql 5432 port, should not have this error, please help check, thanks

**Comment by yinan8128** [comment]
@kmoppel 

After service up 1 weeks 4 days, the error showed 9 times, is there any other info we need to provide please? Thanks

CT-932280701 /# systemctl status pgwatch2 -l
● pgwatch2.service - PGWatch2
   Loaded: loaded (/etc/systemd/system/pgwatch2.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/pgwatch2.service.d
           └─override.conf, required.conf
   Active: active (running) since Thu 2021-04-01 03:51:43 UTC; 1 weeks 4 days ago
 Main PID: 5951 (pgwatch2-daemon)
   CGroup: /system.slice/pgwatch2.service
           └─5951 /usr/bin/pgwatch2-daemon --datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1 --dbname=postgres --password=

Apr 01 03:51:43 s34567y-patroni_cluster-932280701.acronis systemd[1]: Started PGWatch2.
Apr 03 12:36:07 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/03 12:36:07 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 06 09:00:36 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/06 09:00:36 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 07 04:48:44 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/07 04:48:44 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 08 14:02:59 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/08 14:02:59 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 09 17:41:11 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/09 17:41:11 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 10 01:33:14 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/10 01:33:14 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 10 21:15:22 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/10 21:15:22 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 11 04:21:26 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/11 04:21:26 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 11 05:55:26 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/11 05:55:26 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist

**Comment by kmoppel** [comment]
@yinan8128 Thanks for the extra info. Could not spot anything suspicious immediately...but I've now set up a test environment with similar config and will let it run for a week.

**Comment by yinan8128** [comment]
@kmoppel 

Thank you for your update.

Is there any error log in the test environment please?

**Comment by yinan8128** [comment]
@kmoppel 

Any update please? Is there the error log in the test environment?

**Comment by kmoppel** [comment]
@yinan8128 nope, cannot still get any errors after almost a month. Just some reboots it seems
```
root@pgwatch2-long-term-test:~# journalctl -u pgwatch2.service 
-- Logs begin at Thu 2021-04-01 08:36:53 UTC, end at Mon 2021-04-26 09:06:01 UTC. --
Apr 01 10:38:01 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 01 10:38:30 pgwatch2-long-term-test systemd[1]: Stopping Pgwatch2 Gathering Daemon...
Apr 01 10:38:30 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Succeeded.
Apr 01 10:38:30 pgwatch2-long-term-test systemd[1]: Stopped Pgwatch2 Gathering Daemon.
Apr 01 10:38:30 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 01 10:38:31 pgwatch2-long-term-test systemd[1]: Started Pgwatch2 Gathering Daemon.
Apr 01 10:38:33 pgwatch2-long-term-test systemd[1]: Stopping Pgwatch2 Gathering Daemon...
Apr 01 10:38:33 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Succeeded.
Apr 01 10:38:33 pgwatch2-long-term-test systemd[1]: Stopped Pgwatch2 Gathering Daemon.
Apr 01 10:38:33 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 01 10:38:33 pgwatch2-long-term-test systemd[1]: Started Pgwatch2 Gathering Daemon.
Apr 13 07:17:35 pgwatch2-long-term-test pgwatch2-daemon[2395]: 2021/04/13 07:17:35 ERRO GetMonitoredDatabasesFromMonitoringConfig: Failed to resolve DBs for "test": Failed to connect to any>
Apr 13 07:18:15 pgwatch2-long-term-test systemd[1]: Stopping Pgwatch2 Gathering Daemon...
Apr 13 07:18:15 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Succeeded.
Apr 13 07:18:15 pgwatch2-long-term-test systemd[1]: Stopped Pgwatch2 Gathering Daemon.
Apr 13 07:18:15 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 13 07:18:15 pgwatch2-long-term-test systemd[1]: Started Pgwatch2 Gathering Daemon.
Apr 13 07:18:15 pgwatch2-long-term-test pgwatch2-daemon[859221]: 2021/04/13 07:18:15 ERRO GetMonitoredDatabasesFromMonitoringConfig: Failed to resolve DBs for "test": Failed to connect to a>
Apr 13 07:18:15 pgwatch2-long-term-test pgwatch2-daemon[859221]: 2021/04/13 07:18:15 WARN main: host info refreshed, nr. of enabled entries in configuration: 0, nr. of distinct metrics: 62
Apr 13 07:18:19 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Current command vanished from the unit file, execution of the command list won't be resumed.
Apr 13 07:18:21 pgwatch2-long-term-test systemd[1]: Stopping Pgwatch2 Gathering Daemon...
Apr 13 07:18:21 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Succeeded.
Apr 13 07:18:21 pgwatch2-long-term-test systemd[1]: Stopped Pgwatch2 Gathering Daemon.
Apr 13 07:18:21 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 13 07:18:21 pgwatch2-long-term-test systemd[1]: Started Pgwatch2 Gathering Daemon.
Apr 13 07:18:21 pgwatch2-long-term-test pgwatch2-daemon[859270]: 2021/04/13 07:18:21 ERRO GetMonitoredDatabasesFromMonitoringConfig: Failed to resolve DBs for "test": Failed to connect to a>
Apr 13 07:18:21 pgwatch2-long-term-test pgwatch2-daemon[859270]: 2021/04/13 07:18:21 WARN main: host info refreshed, nr. of enabled entries in configuration: 0, nr. of distinct metrics: 62
Apr 13 07:20:08 pgwatch2-long-term-test systemd[1]: Stopping Pgwatch2 Gathering Daemon...
Apr 13 07:20:08 pgwatch2-long-term-test systemd[1]: pgwatch2.service: Succeeded.
Apr 13 07:20:08 pgwatch2-long-term-test systemd[1]: Stopped Pgwatch2 Gathering Daemon.
Apr 13 07:21:17 pgwatch2-long-term-test systemd[1]: Starting Pgwatch2 Gathering Daemon...
Apr 13 07:21:17 pgwatch2-long-term-test systemd[1]: Started Pgwatch2 Gathering Daemon.
```

**Comment by yinan8128** [comment]
@kmoppel 

Maybe because your env is based on #348 ticket, could you please setup a new env as same as this ticket to check please? because we setup 2 times all get the error msg, thanks

I move 2 useful comments of #348 ticket here.

As our pgbouncer is using auto anthentication https://www.cybertec-postgresql.com/en/pgbouncer-authentication-made-easy/ , so for monitored_db record has to use "host config" and exclude the template DBs and include ".*" , else service will have error "ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "resolve": Failed to connect to any of the template DBs: [template1 postgres defaultdb]"

Actually the error "GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist" didn't show once pgwatch2 service start, you can check this ticket and #359 ticket log, error happened after several hours or after several days, could you please monitor the env for a longer time please?

we only config pgwatch2.monitored_db md_port to 6432 and remain pgwatch2 service port 5432

Before pgwatch2.monitored_db md_port using 5432 all works well, only once config to 6432 have "ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist"

**Comment by yinan8128** [comment]
@kmoppel 

Below is the pgwatch2 service log these days, please check, how about your env if config as same as ours please?

CT-932280701 /# systemctl status pgwatch2 -l
● pgwatch2.service - PGWatch2
   Loaded: loaded (/etc/systemd/system/pgwatch2.service; enabled; vendor preset: disabled)                              [13/1824]
  Drop-In: /etc/systemd/system/pgwatch2.service.d
           └─override.conf, required.conf
   Active: active (running) since Thu 2021-04-01 03:51:43 UTC; 1 months 3 days ago
 Main PID: 5951 (pgwatch2-daemon)
   CGroup: /system.slice/pgwatch2.service
           └─5951 /usr/bin/pgwatch2-daemon --datastore=prometheus --prometheus-port=9199 --add-real-dbname=true --host=127.0.0.1
--dbname=postgres --password=9a4839f5f6bcfdab03284948036e31f7

Apr 25 01:43:50 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/25 01:43:50 ERRO GetMonitoredDatabasesFromConf$
gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 27 21:04:19 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/27 21:04:19 ERRO GetMonitoredDatabasesFromConf$
gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 28 17:40:28 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/28 17:40:28 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 29 13:50:36 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/29 13:50:36 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 29 19:34:38 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/29 19:34:38 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 29 21:06:39 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/29 21:06:39 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
Apr 30 07:56:43 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/04/30 07:56:43 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
May 02 14:03:04 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/05/02 14:03:04 ERRO GetMonitoredDatabasesFromConf$gDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
May 03 12:49:12 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/05/03 12:49:12 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist
May 05 02:11:26 s34567y-patroni_cluster-932280701.acronis pgwatch2[5951]: 2021/05/05 02:11:26 ERRO GetMonitoredDatabasesFromConfigDB: Failed to resolve DBs for "patroni": pq: unnamed prepared statement does not exist

**Comment by kmoppel** [comment]
You're right, I missed the pgBouncer transaction pooling mode part. Still no errors though...

**Comment by kmoppel** [comment]
I've also updated the Go modules now, including the Postgres driver, could also be some cornercase driver bug that shows up under certain conditions. So I'd recommend to rebuild the gatherer [manually](https://github.com/cybertec-postgresql/pgwatch2/blob/master/pgwatch2/build_gatherer.sh) and re-try.
